packages/core/src/app/pg-core/db/chain-db.ts

```diff
@@ -1,24 +1,523 @@
+    nodeIndex = getParentIndex(nodeIndex)
+  }
+
+  return siblingIndices
+}
+
+const KEYS = {
+  BLOCKS: new BaseKey('b', ['uint32']),
+  DEPOSITS: new BaseKey('d', ['uint256']),
+  TRANSACTIONS: new BaseKey('t', ['uint32', 'uint256']),
+  TREE_HEIGHTS: new BaseKey('h', ['uint32']),
+  TREE_NODES: new BaseKey('n', ['uint32', 'uint32']),
+  LEAF_INDICES: new BaseKey('l', ['uint32', 'uint256']),
+}
+
+/**
+ * Basic ChainDB implementation that provides a nice interface to the chain
+ * database.
+ */
+export class PGChainDB implements ChainDB {
   /**
    * Creates the wrapper.
    * @param db Database to interact with.
    */
-  constructor(private db: KeyValueStore) {
-    const statePrefix = new BaseKey('s')
-    this.statedb = db.bucket(statePrefix.encode())
+  constructor(private db: KeyValueStore) {}
 
-    const historyPrefix = new BaseKey('h')
-    this.historydb = db.bucket(historyPrefix.encode())
+  /**
+   * Queries transactions in a given range.
+   * @param blockNumber Block to look in.
+   * @param start Start of the range to query.
+   * @param end End of the range to query.
+   * @returns a list of transactions in that range.
+   */
+  public async getTransactions(
+    blockNumber: number,
+    start: BigNum,
+    end: BigNum
+  ): Promise<Transaction[]> {
+    const iterator = this.db.iterator({
+      gte: KEYS.TRANSACTIONS.encode([blockNumber, bnToUint256(start)]),
+      lte: KEYS.TRANSACTIONS.encode([blockNumber, bnToUint256(end)]),
+      values: true,
+    })
+
+    const values = await iterator.values()
+    const transactions = values.map((value) => {
+      return decode(value)
+    })
+
+    return transactions
+  }
+
+  /**
+   * Queries deposits in a given range.
+   * @param start Start of the range to query.
+   * @param end End of the range to query.
+   * @returns a list of deposits for the range.
+   */
+  public async getDeposits(start: BigNum, end: BigNum): Promise<Transaction[]> {
+    const iterator = this.db.iterator({
+      gte: KEYS.DEPOSITS.encode([bnToUint256(start)]),
+      lte: KEYS.DEPOSITS.encode([bnToUint256(end)]),
+    })
+
+    const values = await iterator.values()
+    const deposits = values.map((value) => {
+      return decode(value)
+    })
+
+    return deposits
+  }
+
+  /**
+   * Queries the block hash for a given block number.
+   * @param blockNumber Block number to query.
+   * @returns the hash of the given block.
+   */
+  public async getBlockHash(blockNumber: number): Promise<string> {
+    const key = KEYS.BLOCKS.encode([blockNumber])
+    const value = await this.db.get(key)
+    return value.toString()
+  }
+
+  /**
+   * Queries a specific stored Merkle tree node. Attempts to compute the node
+   * from its children if the node doesn't exist.
+   * @param blockNumber Block number to pull the node from.
+   * @param nodeIndex Index of the specific node to query.
+   * @returns the node or `null` if the node cannot be found.
+   */
+  public async getMerkleTreeNode(
```

Comment:
Here's where things get a little complex - this is a recursive method for seeing if we have a given Merkle tree node. If we don't have the node itself, this method checks if we can generate it from some set of children.